src/test/java/net/snowflake/client/jdbc/cloud/storage/GcmEncryptionProviderTest.java

```diff
@@ -0,0 +1,194 @@
+            null,
+            null);
+    byte[] decryptedPlainText = IOUtils.toByteArray(inputStream);
+    assertArrayEquals(plainText, decryptedPlainText);
+  }
+
+  @Test
+  public void testEncryptAndDecryptStreamWithAad() throws Exception {
+    InputStream plainTextStream = new ByteArrayInputStream(plainText);
+
+    // encrypting
+    InputStream encrypted =
+        GcmEncryptionProvider.encrypt(
+            meta, plainText.length, plainTextStream, encMat, storageClient, cekAad, kekAad);
+    byte[] cipherText = IOUtils.toByteArray(encrypted);
+    verify(storageClient)
+        .addEncryptionMetadata(
+            storageObjectMetadataArgumentCaptor.capture(),
+            matDescArgumentCaptor.capture(),
+            cekIvDataArgumentCaptor.capture(),
+            kekIvDataArgumentCaptor.capture(),
+            encKekArgumentCaptor.capture(),
+            contentLengthArgumentCaptor.capture());
+
+    // decrypting
+    InputStream inputStream =
+        GcmEncryptionProvider.decryptStream(
+            new ByteArrayInputStream(cipherText),
+            Base64.getEncoder().encodeToString(encKekArgumentCaptor.getValue()),
+            Base64.getEncoder().encodeToString(cekIvDataArgumentCaptor.getValue()),
+            Base64.getEncoder().encodeToString(kekIvDataArgumentCaptor.getValue()),
+            encMat,
+            cekAad,
+            kekAad);
+    byte[] decryptedPlainText = IOUtils.toByteArray(inputStream);
+    assertArrayEquals(plainText, decryptedPlainText);
+  }
+
+  @Test
+  public void testEncryptAndDecryptFileWithoutAad() throws Exception {
+    File tempFile = Files.createTempFile("encryption", "").toFile();
+    tempFile.deleteOnExit();
+
+    // encrypting
+    InputStream encrypted =
+        GcmEncryptionProvider.encrypt(
+            meta,
+            plainText.length,
+            new ByteArrayInputStream(plainText),
+            encMat,
+            storageClient,
+            null,
+            null);
+    FileUtils.writeByteArrayToFile(tempFile, IOUtils.toByteArray(encrypted));
+    verify(storageClient)
+        .addEncryptionMetadata(
+            storageObjectMetadataArgumentCaptor.capture(),
+            matDescArgumentCaptor.capture(),
+            cekIvDataArgumentCaptor.capture(),
+            kekIvDataArgumentCaptor.capture(),
+            encKekArgumentCaptor.capture(),
+            contentLengthArgumentCaptor.capture());
+
+    // decrypting
+    GcmEncryptionProvider.decryptFile(
+        tempFile,
+        Base64.getEncoder().encodeToString(encKekArgumentCaptor.getValue()),
+        Base64.getEncoder().encodeToString(cekIvDataArgumentCaptor.getValue()),
+        Base64.getEncoder().encodeToString(kekIvDataArgumentCaptor.getValue()),
+        encMat,
+        null,
+        null);
+    byte[] decryptedCipherText = FileUtils.readFileToByteArray(tempFile);
+    assertArrayEquals(plainText, decryptedCipherText);
+  }
+
+  @Test
+  public void testEncryptAndDecryptFileWithAad() throws Exception {
```

Comment:
the tests are quite similar - can we create some helper methods and reuse code?

Reply:
Done.